core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -249,9 +249,14 @@ public void run() {
           delayedTransport.endBackoff();
           boolean shutdownDelayedTransport = false;
           Runnable runnable = null;
+          // TransportSet as a channel layer class should not call into transport methods while
+          // holding the lock, thus we call hasPendingStreams() outside of the lock.  It will cause
+          // a _benign_ race where the TransportSet may transition to CONNECTING when there is not
```

Comment:
You are right... I lost my touch with delayed transport after not touching it for a while